src/main/java/com/transferwise/openbanking/client/api/registration/WebRegistrationClient.java

```diff
@@ -58,21 +59,17 @@ public ClientRegistrationResponse registerClient(
             request.getBody());
 
         try {
-            ResponseEntity<ClientRegistrationResponse> response = restTemplate.exchange(
-                aspspDetails.getRegistrationUrl(),
-                HttpMethod.POST,
-                request,
-                ClientRegistrationResponse.class);
-
-            log.debug("Received registration response with headers '{}' and body '{}'",
-                response.getHeaders(),
-                response.getBody());
-
-            return response.getBody();
-        } catch (RestClientResponseException e) {
+            var prefixLog = "Received registration response";
+            return webClient.post()
+                .uri(aspspDetails.getRegistrationUrl())
+                .headers(httpHeaders -> httpHeaders.addAll(request.getHeaders()))
+                .bodyValue(Validate.notNull(request.getBody()))
+                .exchangeToMono(clientResponse -> exchangeToMonoWithLog(clientResponse, prefixLog, ClientRegistrationResponse.class))
+                .block();
+        } catch (WebClientResponseException e) {
             throw new ApiCallException("Call to register client endpoint failed, body returned '" + e.getResponseBodyAsString() + "'",
                 e);
-        } catch (RestClientException e) {
+        } catch (WebClientException e) {
```

Comment:
I think we need to change complete flow. This is moving in the direction of asynchronous and non blocking programming (Reactive). 

- We should avoid using block. 
- Return the Mono. 
- try catch could be replaced with onErrorResume. we are using this pisp service. 
- block will done by caller if required otherwise chaining will be used. 

Something like this: 
```
public Mono<ClientRegistrationResponse> registerClient(Request request, AspspDetails aspspDetails) {
        var prefixLog = "Received registration response";

        return webClient.post()
                .uri(aspspDetails.getRegistrationUrl())
                .headers(httpHeaders -> httpHeaders.addAll(request.getHeaders()))
                .bodyValue(Validate.notNull(request.getBody()))
                .exchangeToMono(clientResponse -> exchangeToMonoWithLog(clientResponse, prefixLog, ClientRegistrationResponse.class))
                .doOnSuccess(response -> log.debug("Received registration response with headers '{}' and body '{}'", 
                        response.headers().asHttpHeaders(), response.body()))
                .onErrorResume(WebClientResponseException.class, e -> {
                    log.error("Call to register client endpoint failed, body returned '{}'", e.getResponseBodyAsString(), e);
                    return Mono.error(new ApiCallException("Call to register client endpoint failed, body returned '" + e.getResponseBodyAsString() + "'", e));
                })
                .onErrorResume(WebClientException.class, e -> {
                    log.error("Call to register client endpoint failed, and no response body returned", e);
                    return Mono.error(new ApiCallException("Call to register client endpoint failed, and no response body returned", e));
                });
    }
```

Reply:
We can also do partial change like using block() here for time being. Still worth to remove all the blocking try catch.

Reply:
> We can also do partial change like using block() here for time being.

Yes, I agreed. I just looked at what needs to be changed if we are going to remove `block()`. It involves a lot of files, which I think we better do it after we are done with `webClient` replacement.

> We can also do partial change like using block() here for time being. Still worth to remove all the blocking try catch.

Done!